proposals/0248-tpu-feedback.md

```diff
@@ -0,0 +1,186 @@
+information to more accurately perform QOS. For example, if a peer is
+persistently submitting high percentage pf transactions which are rejected,
+its permitted throughput might be reduced.
+
+## Alternatives Considered
+
+There are different transport mechanism to send the feedback such as
+bidirectional streams, reverse unistream from the server to the client and
+QUIC datagrams. We choose datagrams due to its simplicity. One reason we have
+used stream over datagrams is the expectation that transaction size can be
+increased in the future which can not be carried in one datagram. Sending
+feedback has no such concern as we can send multiple datagrams if there are
+too many transactions to acknowledge.
+
+## New Terminology
+
+In this document we define the following,
+
+TPU Feedback -- information sent by the server about transaction processing.
+Server -- the validator receiving the TPU transactions and sending the feedback.
+Client -- the validator or application using QUIC TPU client sending the TPU
+          transactions and receiving the feedback.
+
+## Detailed Design
+
+When the client sends transactions to the server, it must first open a QUIC
+connection to the server. The feedback is sent by the server using the same
+connection. The feedback is sent using a QUIC datagram. The client wishing to
+receive feedback from the server must be ready to receive application datagrams
+from its connections.
+
+A client connection notifies the server of its interest to receive feedback by
+sending a datagram soliciting feedback. This helps server to avoid sending
+datagrams to clients which are not interested in receiving feedback or clients
+which are not upgraded to have such capabilities.
+
+The client and server must enable the support of datagrams QUIC frame types
+via transport parameter `max_datagram_frame_size`.
+
+The server sends the feedback datagrams to interested clients periodically.
+The frequency is decided by the server and consideration should be given to
+coalesce feedback on more transactions to reduce datagram counts and short
+enough feedback latency. For example 100ms can be considered. If there are
+more transactions to be acknowledged in one datagram, more datagrams
+can be sent without waiting for the feedback interval.
+
+### The format of the TPU feedback
+
+struct TpuFeedback {
+  version: u8,
+  timestamp: u64,
+  transaction_state: TransactionState,
+  priority_fee_info: PriorityFeeInfo
+}
+
+The version is a 8 bit unsinged integer, it is set to 1 for first version. It
+is designed to allow the protocol to extend for future extension.
+
+The timestamp is the 64 bit unsigned integer representing the number of seconds
+since midnight 01/01/1970.
+
+TransactionState is defined as the following:
+
+struct TransactionState {
+  transactions_count: u8
+  (transaction, state) pairs
+}
+
+The transactions_count is 8 bit unsgined integer representing the count of
```

Comment:
Suggest we use TLV encoding here rather than a fixed format. Typically, TLV encodings provide far better ability for upgrade. Also with TLV we do not need to include total count at all.